File: 5-longest-palindromic-substring/longest-palindromic-substring.js
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {
    maxLen = 0
    maxStart = 0
    maxEnd = 0

    function iterateInner(left, right) {
        while(left >= 0 && right < s.length) {
            if(s[left] === s[right]) {
                if(right - left + 1 > maxLen) {
                    maxLen = right - left + 1
                    maxStart = left
                    maxEnd = right
                }
            }
            else break
            left--
            right++
        }
    }

    for(let i=0; i<s.length; i++) {
        left = i-1
        right= i+1
        iterateInner(left, right)
    }
    for(let i=1; i<s.length; i++) {
        left = i-1
        right= i
        iterateInner(left, right)
    }
    return s.substring(maxStart, maxEnd+1)
};

File: 5-longest-palindromic-substring/longest-palindromic-substring.java
class Solution {
    Boolean[][] dp;
    boolean solve(String s, int start, int end) {
        if(start == end) return true;
        if(start > end) return true;
        if(dp[start][end] != null) return dp[start][end];
        boolean ans = false;
        if(s.charAt(start) == s.charAt(end) && solve(s, start+1, end-1)) ans = true;
        solve(s, start+1, end);
        solve(s, start, end-1);
        return dp[start][end] = ans;
    }
    public String longestPalindrome(String s) {
        dp = new Boolean[s.length()][s.length()];
        solve(s, 0, s.length()-1);
        int maxLen = 0, startMax = 0, endMax = 0;
        for(int i=0; i<s.length(); i++) {
            for(int j=s.length()-1; j>=i; j--) {
                if(dp[i][j] != null && dp[i][j] && (j-i+1) > maxLen) {
                    maxLen = j-i+1;
                    startMax = i;
                    endMax = j;
                    break;
                }
            }
        }
        return s.substring(startMax, endMax+1);
    }
}